### PR TITLE
Add puddles / reagent spills

### PR DESCRIPTION
**Pull request by metalgearsloth**
Requires https://github.com/space-wizards/RobustToolbox/pull/1002

Take two.

Puddles are an entity that uses the SnapGrid so there's a max of 1 per tile. Mops and buckets can transfer this entity's solution elsewhere (and the entity is deleted if it's empty).

Puddles will also overflow once they get too full / evaporate if there's minimal solution for a long time (potentially change by design?)

It's also possible to override the re-coloring (e.g. if you want a generic color like vomit), otherwise it'll use the regular chemical re-coloring (which didn't seem to work before?)

Some other stuff also seemed to throw since I started this yonks ago so I updated that as well.

The sprite to be used will also be randomly selected from the relevant .rsi unless overwritten with the spillhelper (e.g. if someone gets slashed you could force a specific type of sprite).

### CODE REVIEW

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+
+            // Smaller than 1m^3 for now but realistically this shouldn't be hit
+            MaxVolume = 1000;
+
+            // Random sprite state set server-side so it's consistent across all clients
+            _spriteComponent = Owner.GetComponent<SpriteComponent>();
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+            var randomVariant = robustRandom.Next(0, _spriteVariants - 1);
+            // TODO: This is ugly
+            string[] splitRSI = _spriteComponent.BaseRSIPath.Split("/".ToCharArray());
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs]: Probably better to use `ResourcePath` to operate on this instead.

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+            }
+            // Opacity based on level of fullness to overflow
+            // Hard-cap lower bound for visibility reasons
+            var volumeScale = ((float)CurrentVolume / OverflowVolume) * 0.75f + 0.25f;
+            var cappedScale = Math.Min(1.0f, volumeScale);
+            // Color based on the underlying solutioncomponent
+            Color newColor;
+            if (_recolor)
+            {
+                newColor = new Color(
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs]: you can do `_contents.SubstanceColor.WithAlpha(cappedScale)` here.

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+                }
+
+                // If we found an existing puddle on that tile then we don't need to spawn a new one
+                var noSpawn = false;
+
+                foreach (var entity in _snapGrid.GetInDir(direction))
+                {
+                    // Don't overflow to walls
+                    if (entity.TryGetComponent(out CollidableComponent collidableComponent) &&
+                        collidableComponent.CollisionLayer == 1)
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs]: Please use collision groups defined in `Content.Shared.Physics.CollisionGroup` instead.

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 4]
```diff
@@ -106,8 +104,7 @@ public override void Initialize()
             _spriteComponent = Owner.GetComponent<SpriteComponent>();
             var robustRandom = IoCManager.Resolve<IRobustRandom>();
             var randomVariant = robustRandom.Next(0, _spriteVariants - 1);
-            // TODO: This is ugly
-            string[] splitRSI = _spriteComponent.BaseRSIPath.Split("/".ToCharArray());
+            string[] splitRSI = _spriteComponent.BaseRSIPath.Split(ResourcePath.SYSTEM_SEPARATOR);
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs]: I meant that more as in `var baseName = new ResourcePath(_spriteComponent.BaseRSIPath).FileName.Replace("rsi", "");`

File: Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs [hunk 5]
```diff
@@ -105,7 +105,7 @@ public override void Initialize()
             var robustRandom = IoCManager.Resolve<IRobustRandom>();
             var randomVariant = robustRandom.Next(0, _spriteVariants - 1);
             string[] splitRSI = _spriteComponent.BaseRSIPath.Split(ResourcePath.SYSTEM_SEPARATOR);
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Fluids/PuddleComponent.cs:107]: You forgot to remove this var
```suggestion
            
```

**Comment by Injazz** [comment]
i've made some changes to chemistry, and i'll do some more in near future. I'll tell when im done!

**Pull request merged**